chain/chaintest/block.go

```diff
@@ -104,3 +111,270 @@ func GenerateTestExecutedBlocks(
+// transactions followed by writing the chain metadata to the state diff.
+func GenerateExecutionBlocks(
+	ctx context.Context,
+	rules chain.Rules,
+	metadataManager chain.MetadataManager,
+	balanceHandler chain.BalanceHandler,
+	txListGenerator TxListGenerator,
+	parentCtx *parentContext,
+	numBlocks uint64,
+	numOfTxsPerBlock uint64,
+) ([]*chain.ExecutionBlock, error) {
+	var timestampOffset int64
+	timestampOffset = rules.GetMinEmptyBlockGap()
+	if numOfTxsPerBlock > 0 {
+		timestampOffset = rules.GetMinBlockGap()
+	}
+	executionBlocks := make([]*chain.ExecutionBlock, numBlocks)
+	for i := range executionBlocks {
+		timestamp := parentCtx.timestamp + timestampOffset*(int64(i)+1)
+		height := parentCtx.height + 1 + uint64(i)
+
+		feeBytes, err := parentCtx.view.GetValue(ctx, chain.FeeKey(metadataManager.FeePrefix()))
+		if err != nil {
+			return nil, err
+		}
+		feeManager := internalfees.NewManager(feeBytes)
+		feeManager = feeManager.ComputeNext(timestamp, rules)
+		unitPrices := feeManager.UnitPrices()
+
+		txGenerator := func(actions []chain.Action, factory chain.AuthFactory) (*chain.Transaction, error) {
+			units, err := chain.EstimateUnits(rules, actions, factory)
+			if err != nil {
+				return nil, err
+			}
+
+			maxFee, err := fees.MulSum(unitPrices, units)
+			if err != nil {
+				return nil, err
+			}
+
+			txData := chain.NewTxData(
+				chain.Base{
+					Timestamp: utils.UnixRMilli(timestamp, rules.GetValidityWindow()),
+					ChainID:   rules.GetChainID(),
+					MaxFee:    maxFee,
+				},
+				actions,
+			)
+
+			return txData.Sign(factory)
+		}
+
+		txs, err := txListGenerator(numOfTxsPerBlock, txGenerator)
+		if err != nil {
+			return nil, err
+		}
+
+		ts := tstate.New(0)
+		for _, tx := range txs {
+			stateKeys, err := tx.StateKeys(balanceHandler)
+			if err != nil {
+				return nil, err
+			}
+			tsv := ts.NewView(stateKeys, parentCtx.view, 0)
+			if err := tx.PreExecute(ctx, feeManager, balanceHandler, rules, tsv, timestamp); err != nil {
+				return nil, err
+			}
+			result, err := tx.Execute(ctx, feeManager, balanceHandler, rules, tsv, timestamp)
+			if err != nil {
+				return nil, err
+			}
+			ok, d := feeManager.Consume(result.Units, rules.GetMaxBlockUnits())
+			if !ok {
+				return nil, fmt.Errorf("consumed too many units in dim %d", d)
+			}
+			tsv.Commit()
+		}
+
+		tsv := ts.NewView(state.CompletePermissions, parentCtx.view, 0)
+		if err := tsv.Insert(ctx, chain.HeightKey(metadataManager.HeightPrefix()), binary.BigEndian.AppendUint64(nil, height)); err != nil {
+			return nil, err
+		}
+		if err := tsv.Insert(ctx, chain.TimestampKey(metadataManager.TimestampPrefix()), binary.BigEndian.AppendUint64(nil, uint64(timestamp))); err != nil {
+			return nil, err
+		}
+		if err := tsv.Insert(ctx, chain.FeeKey(metadataManager.FeePrefix()), feeManager.Bytes()); err != nil {
+			return nil, err
+		}
+		tsv.Commit()
+
+		root, err := parentCtx.view.GetMerkleRoot(ctx)
```

Comment:
Can we rename this `parentRoot` for clarity? I thought this was a bug using the wrong root on first read